taskvine/src/worker/vine_cache.c

```diff
@@ -376,99 +387,163 @@ static int do_transfer( struct vine_cache *c, const char *source_url, const char
 	return result;
 }
 
+/* 
+scan through transfer processes and mark them as completed as they finish.
+*/
+
+int vine_cache_wait(struct vine_cache *c, struct link *manager)
```

Comment:
Please factor this function into several smaller functions that:
- iterate over the table
- process one entry
- evaluate the exit status
- extract the outputs

I think that will make the behavior more transparently clear.